state/src/impls/shard_level.rs

```diff
@@ -411,6 +337,86 @@ impl<B: Backend + ShardBackend> ShardLevelState<B> {
+            OwnedAssetAddress::new(input.prev_out.transaction_hash, input.prev_out.index, self.shard_id);
+        let asset_scheme_address = AssetSchemeAddress::from_hash(input.prev_out.asset_type)
+            .ok_or(TransactionError::AssetSchemeNotFound(input.prev_out.asset_type.into()))?;
+
+        let asset_scheme = self
+            .asset_scheme((&asset_scheme_address).into())?
+            .ok_or(TransactionError::AssetSchemeNotFound(asset_scheme_address.into()))?;
+
+        if let Some(ref registrar) = asset_scheme.registrar() {
+            if registrar != sender {
+                return Err(TransactionError::NotRegistrar(Mismatch {
+                    expected: *registrar,
+                    found: *sender,
+                }).into())
+            }
+        }
+
+        match self.asset(&asset_address)? {
+            Some(asset) => {
+                if asset.amount() != &input.prev_out.amount {
+                    return Err(TransactionError::InvalidAssetAmount {
+                        address: asset_address.into(),
+                        expected: *asset.amount(),
+                        got: input.prev_out.amount,
+                    }.into())
+                }
+                Ok((asset, asset_address))
+            }
+            None => Err(TransactionError::AssetNotFound(asset_address.into()).into()),
+        }
+    }
+
+    fn check_and_run_input_script(
+        &self,
+        input: &AssetTransferInput,
+        transaction_hash: &PartialHashing,
+        cur: &AssetOutPoint,
+        burn: bool,
+    ) -> StateResult<ScriptResult> {
+        let (address_hash, asset) = {
+            let index = input.prev_out.index;
+            let address = OwnedAssetAddress::new(input.prev_out.transaction_hash, index, self.shard_id);
+            match self.asset(&address)? {
+                Some(asset) => (address.into(), asset),
+                None => return Err(TransactionError::AssetNotFound(address.into()).into()),
+            }
+        };
+
+        if *asset.lock_script_hash() != Blake::blake(&input.lock_script) {
+            return Err(TransactionError::ScriptHashMismatch(Mismatch {
+                expected: *asset.lock_script_hash(),
+                found: Blake::blake(&input.lock_script),
+            }).into())
+        }
+
+        let script_result = match (decode(&input.lock_script), decode(&input.unlock_script)) {
+            (Ok(lock_script), Ok(unlock_script)) => execute(
+                &unlock_script,
+                &asset.parameters(),
+                &lock_script,
+                transaction_hash,
+                VMConfig::default(),
+                cur,
+                burn,
+            ),
+            // FIXME : Deliver full decode error
+            _ => return Err(TransactionError::InvalidScript.into()),
+        }.map_err(|err| {
+            ctrace!(TX, "Cannot run unlock/lock script {:?}", err);
+            return TransactionError::FailedToUnlock(address_hash)
```

Comment:
Remove `return`

Reply:
I removed it. Is there any special reason for not to use `return` in function as an argument?

Reply:
It's not because it's a callback function.
It's just because it's not needed. The last expression of the function is returned.